File: Kattis/Equality_Control.cpp
#include<bits/stdc++.h>
using namespace std;

struct Node{
  string type;
  Node *l, *r, *v;
  list<int> ls;

  ~Node(){
    if(l!=nullptr)
      delete l;
    if(r!=nullptr)
      delete r;
    if(v!=nullptr)
      delete v;
  }

};

Node* shuf(Node *v) {
  return new Node{"shuf", nullptr, nullptr, move(v)};
}

Node* sort(Node *v) {
  return new Node{"sort", nullptr, nullptr, move(v)};
}

Node* ls(list<int> v) {
  return new Node{"list", nullptr, nullptr, nullptr, move(v)};
}

Node* cat(Node *l, Node *r) {
  return new Node{"cat", l, r};
}

bool allEq(list<int> &l){
  for(int v : l)
    if(v!=l.front())
      return false;
  return true;
}

Node* parse(string &s) {
  int i = 0;
  stack<string> fst;
  stack<Node*> vst;
  while(i < s.size()){
    if(s[i] >= 'a' && s[i] <= 'z'){
      string o = "";
      while(s[i] >= 'a' && s[i] <= 'z'){
        o+=s[i];
        i+=1;
      }
      i+=1; //consume (
      fst.push(o);
    }else if(s[i] == ')') {
      i+=1;
      string c = fst.top();fst.pop();

      if(c=="concat"){
        auto r = vst.top();vst.pop();
        auto l = vst.top();vst.pop();

        vst.push(cat(l,r));
      }else if(c=="shuffle"){

        auto v = vst.top();vst.pop();

        if(v->type=="list" && allEq(v->ls))
          vst.push(v);
        else
          vst.push(shuf(v));
      }else if(c=="sorted"){
        auto v = vst.top(); vst.pop();
        vst.push(sort(v));
      }
    } else if(s[i] == '['){
      ++i;
      list<int> o;
      while(s[i] != ']'){
        int a = 0;
        while(s[i] >= '0' && s[i] <= '9'){
          a = 10*a + (s[i]-'0');
          ++i;
        }
        if(a)
          o.push_back(a);
        if(s[i]==',')
          ++i;
      }
      vst.push(ls(o));
    } else {
      i+=1;
    }
  }
  assert(vst.size()==1);
  return vst.top();
}

Node* pr(Node * t){
  cout << t->type << endl;
  if(t->l)
    pr(t->l);
  if(t->r)
    pr(t->r);
  if(t->v)
    pr(t->v);
  if(!t->ls.empty()){
    for(int v : t->ls)
      cout << v << ',';
    cout << endl;
  }
  return t;
}

Node* comp(Node* t){
  stack<Node*> st;
  st.push(t);
  stack<Node* > vst;
  while(!st.empty()){
    auto v = st.top();
    if(v->type == "list"){
      vst.push(v); //osort?
      st.pop();
    }
    else if(v->type=="cat"){
      if(v->l != nullptr){
        st.push(v->l);
        v->l = nullptr;
      }
      else if(v->r != nullptr){
        st.push(v->r);
        v->r=nullptr;
      }
      else{
        auto r = vst.top(); vst.pop();

        vst.top()->ls.splice(vst.top()->ls.end(), r->ls);

        st.pop();
      }
    }
    else {
      auto r = v->v;
      st.pop();
      st.push(r);
    }
  }

  assert(vst.size()==1);
  return vst.top();
}

Node* osort(Node *v){
  if(v->type != "list")
    return v;
  v->ls.sort();
  return v;
}

Node* ev(Node * v) {
  stack<Node*> st;
  st.push(v);
  stack<Node*> vst;
  while(!st.empty()){
    auto t = st.top();
    string n = t->type;
    if(n == "list"){
      vst.push(t);
      st.pop();
    }
    if(n == "cat"){
      if(t->l != nullptr){
        st.push(t->l);
        t->l=nullptr;
      }else if(t->r != nullptr){
        st.push(t->r);
        t->r=nullptr;
      } else {
        auto r = vst.top();vst.pop();
        auto l = vst.top();vst.pop();
        vst.push(cat(l, r));
        st.pop();
      }
    }
    if(n == "shuf"){
      vst.push(shuf(comp(t->v)));
      st.pop();
    }
    if(n == "sort"){
      vst.push(osort(comp(t->v)));
      st.pop();
    }
  }
  assert(vst.size()==1);
  return vst.top();
}

list<int> gen(Node* v) {
  stack<Node*> st;
  st.push(v);
  list<int> l;
  while(!st.empty()){
    auto t = st.top();
    string n = t->type;
    if(n == "list"){
      l.splice(l.end(), t->ls);
      st.pop();
    }
    else if(n=="cat"){
      if(t->l != nullptr){
        st.push(t->l);
        t->l = nullptr;
      } else if(t->r != nullptr){
        auto r = t->r;
        st.pop();
        st.push(r);
      }
    }
    else if(n=="shuf") {
      l.push_back(-1);
      l.splice(l.end(), osort(t->v)->ls);
      l.push_back(-1);
      st.pop();
    }
  }
  return l;
}

int main(){
  string s1,s2;
  getline(cin,s1);
  getline(cin,s2);
  auto p = gen(ev(parse(s1)));
  auto q = gen(ev(parse(s2)));
  if(p==q)
    puts("equal");
  else
    puts("not equal");
  return 0;
}

File: Kattis/Equality_Control.py
import time
import sys
osort = sorted

class cat:
  def __init__(self,l,r):
    self.l=l
    self.r=r
    self.t = 'cat'

class sort:
  def __init__(self,v):
    self.v=v
    self.t = 'sort'


class shuf:
  def __init__(self, v):
    self.v=v
    self.t = 'shuf'

def alleq(xs):
  for x in xs:
    if x != xs[0]:
      return False
  return True

def parse(s):
  i = 0
  fst = []
  vst = []
  while i < len(s):
    if s[i] >= 'a' and s[i] <= 'z':
      o = ""
      while s[i] >= 'a' and s[i] <= 'z':
        o+=s[i]
        i+=1
      i+=1 #consume (
      fst.append(o)

    elif s[i] == ')':
      i+=1
      c = fst.pop()
      if c=='concat':
        r = vst.pop()
        l = vst.pop()

        vst.append(cat(l,r))
      elif c=='shuffle':
        v = vst.pop()
        if isinstance(v,(list,)) and alleq(v):
          vst.append(v)
        else:
          vst.append(shuf(v))
      elif c=='sorted':
        v = vst.pop()
        vst.append(sort(v))

    elif s[i] == '[':
      o = ""
      while s[i] != ']':
        o+=s[i]
        i+=1
      o+=s[i]
      i+=1
      vst.append(eval(o))

    elif s[i] == ',':
      i+=1

  assert(len(vst)==1)

  return vst[0]

'''
def comp(t):
  if isinstance(t, (list,)):
    return osort(t)
  if type(t).__name__ == 'cat':
    return comp(t.l) + comp(t.r)
  return comp(t.v)
'''

def comp(t):
  cs = time.time()
  st = [t]
  vst = []
  cm = 0
  while len(st):
    v = st[-1]
    if isinstance(v, (list,)):
      vst.append(osort(v))
      st.pop()
    elif v.t == 'cat':
      if v.l != None:
        st.append(v.l)
        st[-2].l = None
      elif v.r != None:
        st.append(v.r)
        st[-2].r = None
      else:
        r = vst.pop()

        tsts = time.time()
        vst[-1]+=r
        tste = time.time()

        cm += tste-tsts

        st.pop()
    else:
      st.pop()
      st.append(v.v)

  assert(len(vst)==1)
  ce = time.time()
  print "comp took", ce-cs
  print "catting took", cm
  return vst[0]

'''
def ev(t):
  if isinstance(t, (list,)):
    return t
  n = type(t).__name__
  if n=='cat':
    return cat(ev(t.l), ev(t.r))
  if n=='shuf':
    return shuf(comp(t.v))
  if n=='sort':
    return osort(comp(t.v))

'''

def ev(v):
  st = [v]
  vst = []
  while len(st):
    t = st[-1]
    if isinstance(t, (list,)):
      vst.append(t)
      st.pop()
      continue
    n = t.t
    if n == 'cat':
      if t.l != None:
        st.append(t.l)
        st[-2].l=None
      elif t.r != None:
        st.append(t.r)
        st[-2].r=None
      else:
        r = vst.pop()
        l = vst.pop()
        vst.append(cat(l, r))
        st.pop()
    if n == 'shuf':
      vst.append(shuf(comp(t.v)))
      st.pop()
    if n == 'sort':
      vst.append(osort(comp(t.v)))
      st.pop()

  assert(len(vst)==1)
  return vst[0]

'''
def gen(t):
  if isinstance(t, (list,)):
    return t
  if type(t).__name__ == 'cat':
    return gen(t.l) + gen(t.r)

  if type(t).__name__ == 'shuf':
    return [-1] + t.v + [-1]
'''

def gen(v):
  st = [v]
  l = []
  while len(st):
    t = st[-1]
    if isinstance(t, (list,)):
      l += t
      st.pop()
    elif t.t=='cat':
      if t.l != None:
        st.append(t.l)
        st[-2].l = None
      elif t.r != None:
        st.pop()
        st.append(t.r)
    elif t.t=='shuf':
      l.append(-1)
      l += osort(t.v)
      l.append(-1)
      st.pop()

  return str(l)

'''
print gen(ev(parse(raw_input())))
print gen(ev(parse(raw_input())))
'''

if __name__ == "__main__":
  s1 = raw_input()
  s2 = raw_input()
  ps = time.time()
  p = parse(s1)
  pe = time.time()
  print "parse took", pe-ps
  es = time.time()
  e = ev(p)
  ee = time.time()
  print "ev took", ee-es
  gs = time.time()
  g = gen(e)
  ge = time.time()
  print "gen took", ge-gs
  if g == gen(ev(parse(s2))):
    print "equal"
  else:
    print "not equal"
